examples/sugarscape_ig/ss_polars/agents.py

```diff
@@ -104,22 +147,160 @@ def move(self):
+            )
+            .drop("agent_order")
         )
+        return best_moves
 
-    def step(self):
-        self.shuffle().do("move").do("eat")
-        self.discard(self.agents.filter(pl.col("sugar") <= 0))
+    def _prepare_cells(self, neighborhood: pl.DataFrame):
+        occupied_cells = (
+            neighborhood[["agent_id_center"]]
+            .unique()
+            .join(self.pos, left_on="agent_id_center", right_on="unique_id")
+            .with_columns(
+                flattened=(pl.col("dim_0") * self.space.dimensions[1] + pl.col("dim_1"))
+            )["flattened"]
+            .to_numpy()
+        )
+        free_cells = np.ones(
+            self.space.dimensions[0] * self.space.dimensions[1], dtype=np.bool_
+        )
+        free_cells[occupied_cells] = False
+        target_cells = (
+            neighborhood["dim_0"] * self.space.dimensions[1] + neighborhood["dim_1"]
+        ).to_numpy()
+        return occupied_cells, free_cells, target_cells
+
+    def _get_best_moves(self):
+        raise NotImplementedError("Subclasses must implement this method")
+
+
+class AntPolarsLoopNoVec(AntPolarsLoop):
+    def _get_best_moves(self):
+        def inner_get_best_moves(
+            occupied_cells: np.ndarray,
+            free_cells: np.ndarray,
+            target_cells: np.ndarray,
+            agent_id_center: np.ndarray,
+            blocking_agent: np.ndarray,
+            processed_agents: np.ndarray,
+            best_moves: np.ndarray,
+        ) -> np.ndarray:
+            for i, agent in enumerate(agent_id_center):
+                if not processed_agents[agent]:
+                    if free_cells[target_cells[i]] or blocking_agent[i] == agent:
+                        best_moves[agent] = target_cells[i]
+                        # Free current cell
+                        free_cells[occupied_cells[agent]] = True
+                        # Occupy target cell
+                        free_cells[target_cells[i]] = False
+                        processed_agents[agent] = True
+            return best_moves
+
+        return inner_get_best_moves
+
+
+class AntPolarsNumba(AntPolarsLoop):
+    def _get_best_moves(self):
+        @guvectorize(
+            [
+                (
+                    int64[:],
+                    int64[:],
+                    int64[:],
+                    int64[:],
+                    int64[:],
+                    int64[:],
+                    int64[:],
+                )
+            ],
+            "(n), (m), (p), (p), (p), (n)->(n)",
+            nopython=True,
+            target=self.numba_target,
+        )
+        def vectorized_get_best_moves(
+            occupied_cells,
+            free_cells,
+            target_cells,
+            agent_id_center,
+            blocking_agent,
+            processed_agents,
+            best_moves,
+        ):
+            for i, agent in enumerate(agent_id_center):
```

Comment:
It seems that the Numba constraints (only 1D, fixed sized vectors, input vector) don't compromise its readability that much. I find it a worth-it contender.

Reply:
Yeah, it's not that bad. It's just the conversion from 2d grid to 1d grid which is a bit "hacky". I will solve the cuda problem so we can merge into main and solve #